Twister: runner: Option to filter test/sample projects post build

tychofrei02 opened the pull request:
Added a the possibility to only run test/sample projects if the source code of their ninja dependencies has changed in relation to the merge base with the main branch. This has to happen post build, when the ninja dependencies are available.

This is an implementation that addresses the following issue: #64313

DeHess committed:
twister: environment: add command line option "--dyn-filter"

Filtering tests and sample projects should be optional,
therefore a new command line option is needed. At the moment, there is
only one way of filtering. In the future, it is possible to add other
choices to filter in different ways.

Signed-off-by: nathan hess <nhess@baumer.com>

tychofrei02 committed:
twister: runner: add method to check for changed dependencies

Using the ninja dependencies available after building and comparing them
with the changed source code relative to the merge base with the main,
the new method "dependencies_have_changed" checks if the current test
instance has any changed dependencies. In this fashion one does not
always have to run every test/sample project, but can skip those,
that are not affected by changes.

Signed-off-by: frei tycho <tfrei@baumer.com>

DeHess committed:
twister: runner: skip tests/sample if their dependencies are unchanged

If the --dyn-filter commandline option is set, tests and sample projects
are checked for changed dependencies. If the dependencies were changed,
the test run is skipped. Additionally, the test / sample instance is
also flagged as skipped.

Signed-off-by: nathan hess <nhess@baumer.com>

DeHess commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1084,6 +1099,25 @@ def run(self):
+    def dependencies_have_changed(self):
+        if self.options.dyn_filter == "changed_src":
+            has_changed_dependencies = False
+            ninja_deps = subprocess.run(['ninja', '-t', 'deps'], capture_output=True, cwd=self.instance.build_dir)
```
The files that you mentioned are not considered with the current approach. The goal of this PR is to offer an option to reduce CI load for situations where only c and h files are touched. We will update the help text to offer more clarity.

hakehuang replied:
I think we can adding support guradually, and need document this limitation and take it as future plan.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -657,23 +661,34 @@ def process(self, pipeline, done, message, lock, results):
-                # to make it work with pickle
-                self.instance.handler.thread = None
-                self.instance.handler.duts = None
+            if self.options.dyn_filter and not self.dependencies_have_changed():
```
Why are you doing this when running? If the purpse of this is static analysis you will be running in --build-only anyways, so you want to do the filtering very early on and not after you have built it, otherwise the cost will be the same.

DeHess replied:
This method is not only intended for static code analysis, and aims to reduce the runtime of long lasting test runs. (For example when running tests on a simulated board.)

If there is a way of getting the ninja deps without actually building, we could and would filter out tests earlier.

tychofrei02 commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -46,6 +46,10 @@
+ZEPHYR_BASE = os.getenv("ZEPHYR_BASE")
+if not ZEPHYR_BASE:
+    sys.exit("$ZEPHYR_BASE environment variable undefined")
+
```
ZEPHYR_BASE is now imported from twisterlib.environment.

tychofrei02 commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1084,6 +1099,25 @@ def run(self):
+            dependencies = [dep.strip() for dep in split_dependencies_list]
+
+            git_diff = subprocess.run(['git', 'diff', '--name-only', '--merge-base', 'main'], capture_output=True)
+            for path in git_diff.stdout.decode().split('\n'):
```
Returncode is now being checked

tychofrei02 committed:
tests: twister: test_runner: mocked dyn_filter option as list

The tests didn't run as expected for the dyn_filter command line option
because dyn_filter is a list of arguments. Since the mock to test it
is automatically created it is not iterable. Now the dyn_filter option
is initialized with a dummy list.

Signed-off-by: frei tycho <tfrei@baumer.com>

DeHess commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -1084,6 +1099,25 @@ def run(self):
         sys.stdout.flush()
 
+    def dependencies_have_changed(self):
+        if self.options.dyn_filter == "changed_src":
```
We have now made sure that the --dyn-filter option is mocked correctly as an iterable

tychofrei02 commented:
All requested changes have been addressed. Please re-review.

hakehuang commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -108,6 +108,13 @@ def add_parse_arguments(parser = None):
              "invocation.")
 
+    case_select.add_argument(
+        "--dyn-filter",
```
need update to ducoment as well, and need add a simple command example on how to use it

romkell replied:
@hakehuang 
> need update to ducoment as well, and need add a simple command example on how to use it

You are right, needs more and better description for the user.

Idea is to reduce the CI load by only running tests on projects which have changed sources (compared to the git merge-base) and not every project.

At the moment the implementation needs to build all projects (to get the .ninja-deps file to make the association between changes sources and project) and then only executes the tests for those that have changed sources.

That only makes sense, where the build is much faster than the tests e.g. when simulating the MCU.
It also seems not complete. Any non-source project files such as prj.conf etc. are current not considered.

Furthermore the implementation could / should be improved by using the compile_commands.json (instead of ninja-deps).
- Improvement: no full build required as with .ninja-deps, compile_commands.json is created during cmake config phase
- Short-coming: (AFAIKT) there are build time generated sources which are not part of the compile_commands.json.
With using the compile_commands.json, which is created in the configuration phase of cmake, also the builds and not only the tests could be reduced.

For (long lasting) static code analysis (SCA) builds, when using ninja-deps, still a 1. stage build is required and a 2. stage build does the SCA. When using the compile_commands.json, the 1. stage build would not be necessary.

We tried to push the idea of that twister extension, but we also felt the reactions were rather reluctant.

We currently do not know whether or how we proceed.

hakehuang replied:
let's discuss this in test working group meeting, and feedback to you. we know this does not help the current CI, but using dependency is a good step forward. @romkell

hakehuang reviewed (commented):
I am not clear on how to use this, is there any guide?

kartben commented:
rebased to trigger stuck CI jobs
